### ISSUE TITLE
Adding GA script tag?

### ISSUE THREAD

**Issue by robinvdvleuten** [zeit/next.js#160]
Hi there!

I am currently converting my website to a next.js project and can't figure out on how to add the Google Analytics script tag to the page. Any thoughts?

Cheers,
Robin

**Comment by nodegin** [comment]
checkout `<Head>`

**Comment by paulthealien** [comment]
@robinvdvleuten I'd suggest creating a GoogleTagManager component and including it in the Head or right after it depending on your implementation choice.

**Comment by impronunciable** [comment]
as people mentioned you can use the `Head` tag or a react-friendly solution like https://github.com/react-ga/react-ga

**Comment by gtramontina** [comment]
Any tips on getting [react-ga](https://github.com/react-ga/react-ga) working with next.js?

**Comment by impronunciable** [comment]
@gtramontina it should work. Otherwise I'm one of the `react-ga` contributors so I could help with that

**Comment by musemind** [comment]
No need for `react-ga`. Use the google tag manager with the `<Head>` component to inject analytics. There are some small changes to the google tag manager code necessary to make it work in nextjs/react:

```
<Head>
      <script dangerouslySetInnerHTML={{__html: `(function(w,d,s,l,i){w[l]=w[l]||[];w[l].push({'gtm.start':
        new Date().getTime(),event:'gtm.js'});var f=d.getElementsByTagName(s)[0],
        j=d.createElement(s),dl=l!='dataLayer'?'&l='+l:'';j.async=true;j.src=
        'https://www.googletagmanager.com/gtm.js?id='+i+dl;f.parentNode.insertBefore(j,f);
      })(window,document,'script','dataLayer','GTM-XXXXXX');`}} />
    </Head>
    <noscript dangerouslySetInnerHTML={{__html: `<iframe src="https://www.googletagmanager.com/ns.html?id=GTM-XXXXXXX" height="0" width="0" style="display:none;visibility:hidden;"></iframe>`}} />
```

![image](https://cloud.githubusercontent.com/assets/11320794/21155786/c2d09ea6-c173-11e6-8565-0b8578deb831.png)

**Comment by MoOx** [comment]
I would recommend a better solution to track client side page view: autotrack. See https://github.com/MoOx/phenomic/issues/384 for more information.

**Comment by sedubois** [comment]
@impronunciable it would be great to have an example as after some googling I'm still confused 😃 Could this be reopened?

- @MoOx seems to imply that react-ga isn't necessary, should use [Google AutoTrack](https://github.com/googleanalytics/autotrack) instead
- but Autotrack instructs to create a script tag with code which accesses the `window` object, so that doesn't work server-side
- and anyway if using a react component such as react-ga, where should it be put? In a `render`?  `componentWillMount`? `getInitialProps`? It's not super clear as it kind of supposes that you use react-router and load the code a single time for all pages.

The page tracking should work both when the page is server-rendered and then when navigating to another page. It would be would be great to see how the whole thing should integrate in a typical Next app with a Head, a page layout HOC, etc.

Would also be great to know if you think there's a better/simpler/less cumbersome/more state-of-the-art combination than Next.js + Google Analytics Autotrack...

**Comment by impronunciable** [comment]
I can investigate this

**Comment by MoOx** [comment]
Pretty sure autotrack is more than enough, only injected on the client side - no need to think about the server-side.

**Comment by MoOx** [comment]
See https://github.com/MoOx/react-toulouse/commit/c42045d1001ab813c5d7eae5ab2f4f8c08c0025e for an example. It's not using next, but I guess it should be easy to adapt.

**Comment by timneutkens** [comment]
Yeah we don't want to track server renders, Since it Will obfuscate things like location.

**Comment by MoOx** [comment]
No you don't since it will be handled by the client: keep in mind that a page is request by the client and will be rendered in the client browser :)

**Comment by sedubois** [comment]
OK I didn't realise the code shouln't run server-side. So I used autotrack as recommended by @MoOx, I think this should work: https://github.com/relatenow/relate/commit/50dc3f317eb3efa045d5bbe40d1a1a1ad1116458

Just haven't confirmed yet if it counts page views correctly, this is new to me 😬 😄

**Comment by MoOx** [comment]
You can use "real time" view to debug GA easily.

**Comment by sedubois** [comment]
@MoOx yes well that's what bothers me, I don't see anything in the realtime panel when I visit https://relate.now.sh in parallel in another tab (0 active users)

**Comment by sedubois** [comment]
I'm still not getting this... I don't get any data with the code I wrote. 

All the examples I see seem to require a `<script href="https://www.google-analytics.com/analytics.js" async />` but things in the Head don't seem to be executed in the client? (my console.log aren't printed in browser console...)

An example would still be very appreciated 🙂

**Comment by sedubois** [comment]
Ah, got it working by switching to react-ga 😄 

https://github.com/relatenow/relate/commit/a9d2bce46e5314075af7c12bbaa0266865ff25da

<img width="915" alt="screen shot 2017-01-18 at 14 01 28" src="https://cloud.githubusercontent.com/assets/4217871/22063290/bb91e706-dd86-11e6-8d75-d091385df928.png">

**Comment by impronunciable** [comment]
Cool, is it working @sedubois ? time to close this issue for now?

**Comment by luandro** [comment]
How did you get it to work?

I've been trying to call a [pageview](https://github.com/Contemplecom/cg-app/blob/master/utils/analytics.js) function on [componentDidMount](https://github.com/Contemplecom/cg-app/blob/master/pages/index.js#L75)


The production app [GA Check](http://www.gacheck.com/check/http%253A%252F%252Fconsultoriacg.com.br%252F) fails.

**Comment by sedubois** [comment]
@luandro don't remember the details, but you can look at the commit I linked above if it helps

**Comment by luandro** [comment]
I did and tried it. I did the same as you, calling [configureAnalytics](https://github.com/Contemplecom/cg-app/blob/master/utils/analytics.js) on top, and [pageview](https://github.com/Contemplecom/cg-app/blob/master/pages/index.js#L76) on `componentWillMount`, but still get negative on [GA Check](http://www.gacheck.com/check/http%253A%252F%252Fconsultoriacg.com.br%252F). 

In your code  it doesn't seem you actually connected [page hoc](https://github.com/relatenow/relate/blob/a9d2bce46e5314075af7c12bbaa0266865ff25da/hocs/page.js) anywhere, did you?

**Comment by durk0** [comment]
@luandro Simply add `ga(‘set’, ‘page’, window.location.pathname); ga(‘send’, ‘pageview’);` to change url of pageview event.

**Comment by ryardley** [comment]
Maybe there is a more 'next.js' way to do this but WRT running scripts on the page I found that if I created a `Script` component like so:

```js
// modules/Script.js
export default ({children}) => (
  <script dangerouslySetInnerHTML={{__html: `(${children.toString()})();` }}></script>
);
``` 

Then I could do this which worked for me:

```jsx
import Document, { Head, Main, NextScript } from 'next/document'
import Script from '../modules/Script';

export default class MyDocument extends Document {

  render () {
    return (
      <html>
        <body>
          <Main />
          <Script>
            {
              function foo(){
                console.log('foo');
              }
            }
          </Script>
          <NextScript />
        </body>
      </html>
    )
  }
}

```

**Comment by karthikiyengar** [comment]
@luandro - Did you manage to figure this out? 

@durk0 - Could you please elaborate on how the HoC fits into the picture here. It doesn't really seem to be connected.

I've been spending quite some time trying to figure this out using the custom document, but can't really get a reliable way to hook into the location changes. Where exactly do you place the autotrack snippet?

**Comment by karthikiyengar** [comment]
Is it possible to track individual pageviews without explicitly calling the ga('send', 'pageview') on every page. If we absolutely have to do it, where do we invoke it from having certainty that ga has been loaded?

**Comment by vinaypuppal** [comment]
@karthikiyengar I am using `react-ga` as suggested by others above and logging pageviews on `componentDidMount` of every page.
```
import ReactGA from 'react-ga'

export const initGA = () => {
  console.log('GA init')
  ReactGA.initialize('UA-xxxxxxxx-x')
}
export const logPageView = () => {
  ReactGA.set({ page: window.location.pathname })
  ReactGA.pageview(window.location.pathname)
}
```

```
```

**Comment by rongierlach** [comment]
@vinaypuppal is it common practice to initGA for every page component?

**Comment by notrab** [comment]
@rongierlach I'm using a `Layout` component which calls the `initialize` and `pageview` methods on `ReactGA` within `componentDidMount()`.

```js
export default class extends Component {
  componentDidMount() {
    ReactGA.initialize('UA-5341402-1')
    ReactGA.pageview(document.location.pathname)
  }

  render() {
    return (
      <div>
        {this.props.children}
     </div>
    )
}
```

This is the cleanest way I've done it yet as if the page is server rendered, it needs to initialize. I'm sure there is a hook that could bypass the initialize if it's client rendered.

**Comment by kylewiedman** [comment]
@notrab couldn't you potentially be missing the shallow renders?  They your app will just transition and your layout component may not remount.  This is what I have to catch those.

``` js
componentDidMount() {
    ReactGA.initialize('xx-xxxxxxx-x')
    let trackMe = true

    if (trackMe) {
      ReactGA.pageview(document.location.pathname)
      trackMe = false
    }

    Router.onRouteChangeStart = () => {
      NProgress.start()
      this.store.dispatch(transitionPage(true))
      trackMe = true
    }
    Router.onRouteChangeComplete = () => {
      NProgress.done()
      this.store.dispatch(transitionPage(false))
      if (trackMe) { ReactGA.pageview(document.location.pathname) }
    }
    Router.onRouteChangeError = () => {
      NProgress.done()
      this.store.dispatch(transitionPage(false))
    }

    this.store.dispatch(calculateResponsiveState(window))
  }
```

**Comment by filostrato** [comment]
Trying to get this to work, but still no response from Google Analytics. This is my code:

`import ReactGA from 'react-ga'

const dev = process.env.NODE_ENV !== 'production'

export const initGA = () => {
  ReactGA.initialize("UA-100779764-1", {
    debug: dev,
  })
}

export const trackPageView = (
  pageName = window.location.pathname + window.location.search
) => {
  ReactGA.set({page: pageName})
  ReactGA.pageview(pageName)
}

export const trackCustomEvent = (category, action) =>
  ReactGA.event({category, action})

export default undefined`

and in my index page:

`import {initGA, trackPageView} from '../modules/react-ga.js'
[...]
Index.componentDidMount = function() {
	initGA()
	trackPageView()
}`

**Comment by exogenesys** [comment]
@filostrato How did you fix it? I'm stuck at it as well. :)

**Comment by filostrato** [comment]
@exogenesys Ended up putting this in `utils/analytics.js`:

```
import ReactGA from 'react-ga'

const dev = process.env.NODE_ENV !== 'production'

export const initGA = () => {
  ReactGA.initialize("UA-xxxxxxxxx-x", {
    debug: dev,
  })
}

export const logPageView = (
  pageName = window.location.pathname + window.location.search
) => {
  ReactGA.set({page: pageName})
  ReactGA.pageview(pageName)
}

export const trackCustomEvent = (category, action) =>
  ReactGA.event({category, action})

export default undefined
```

and changed my `Layout.js` to extend `React.Component`:

```
export default class Layout extends React.Component {
    componentDidMount () {
      if (!window.GA_INITIALIZED) {
        initGA()
        window.GA_INITIALIZED = true
    }
    logPageView()
  }

  render () {
    return (
      <Main>
        <Header />
        {this.props.children}
      </Main>
    )
  }
}
```

Didn't find a way to make it work in a component which utilized `getInitialProps`, but since `Layout.js` doesn't it worked just fine; not sure what to do if I want more detailed statistics, but I'll cross that bridge when I come to it.

**Comment by exogenesys** [comment]
@filostrato That did it. Thanks a lot. :)

**Comment by osartun** [comment]
Hey guys

I've just dealed with the implementation of GA on my site as well and this is what I've come up with.

First I used the solution from @notrab that seemed clean and straight forward:

```
export default class extends Component {
  componentDidMount() {
    ReactGA.initialize('UA-XXXXXXX-X')
    ReactGA.pageview(document.location.pathname)
  }

  render() {
    …
  }
}
```

However, I noticed:

* `componentDidMount` was executed when I switched between pages of different types, so the `initialize` function was called several times
* `componentDidMount` was _only_ executed when I switched between pages of different types, so the `pageview` function was called only then => I have a `video` pagetype on my website and I saw in the GA Live view that while switching between different video pages only the initial video page was tracked

I created a higher order component to address these two problems. For the first one I used @filostrato's solution.

```
import React, { Component } from 'react';
import ReactGA from 'react-ga';
import Router from 'next/router';

const debug = process.env.NODE_ENV !== 'production';

export default (WrappedComponent) => (
  class GaWrapper extends Component {
    constructor (props) {
      super(props);
      this.trackPageview = this.trackPageview.bind(this);
    }

    componentDidMount() {
      this.initGa();
      this.trackPageview();
      Router.router.events.on('routeChangeComplete', this.trackPageview);
    }

    componentWillUnmount() {
      Router.router.events.off('routeChangeComplete', this.trackPageview);
    }

    trackPageview (path = document.location.pathname) {
      if (path !== this.lastTrackedPath) {
        ReactGA.pageview(path);
        this.lastTrackedPath = path;
      }
    }

    initGa () {
      if (!window.GA_INITIALIZED) {
        ReactGA.initialize('UA-7297359-10', { debug });
        window.GA_INITIALIZED = true;
      }
    }

    render() {
      return (
        <WrappedComponent {...this.props} />
      );
    }
  }
);
```
And in your `<Layout>` / `<PageWrapper>` / `<PageScaffold>` component or whatever you call it:

```
import GaWrapper from './ga-wrapper';

const PageScaffold = () => (…);

export default GaWrapper(PageScaffold);
```

Maybe that helps anyone.

I don't like the call of `Router.router.events.…` though, because this isn't part of the documented API. However, when I tried to call `Router.onRouteChangeComplete` which [should be the correct way](https://github.com/zeit/next.js#router-events), I got a ReferenceError because `onRouteChangeComplete` was `undefined`. Maybe the documentation is outdated?

**Comment by toolgirl** [comment]
This has worked for me in a next.js website. And got around the 'window not defined' error I was running into with the usual implementation of react-ga.

**Comment by florinpop17** [comment]
@osartun solution worked for me. Thank you! 👍

**Comment by jesstelford** [comment]
Thanks to @osartun's code, I was able to come up with the following solution:

`scaffold.js` / `layout.js` / `wraper.js` / whatever your project names it:
```javascript
import url from 'url';

const GA_TRACKING_ID = '...';

const withPageViews = WrappedComponent =>
  class GaWrapper extends React.Component {
    componentDidMount() {
      // We want to do this code _once_ after the component has successfully
      // mounted in the browser only, so we use a special semiphore here.
      if (window.__NEXT_ROUTER_PAGEVIEW_REGISTERED__) {
        return;
      }

      window.__NEXT_ROUTER_PAGEVIEW_REGISTERED__ = true;
      let lastTrackedUrl = '';

      // NOTE: No corresponding `off` as we want this event listener to exist
      // for the entire lifecycle of the page
      // NOTE: This does _not_ fire on first page load. This is what we want
      // since GA already tracks a page view when the tag is first loaded.
      Router.router.events.on('routeChangeComplete', (newUrl = document.location) => {
        if (newUrl === lastTrackedUrl || !window.gtag) {
          return;
        }

        // Don't double track the same URL
        lastTrackedUrl = newUrl;

        // Believe it or not, this triggers a new pageview event!
        // https://developers.google.com/analytics/devguides/collection/gtagjs/single-page-applications
        window.gtag('config', GA_TRACKING_ID, {
          page_path: url.parse(newUrl).path,
        });
      });
    }

    render() {
      return <WrappedComponent {...this.props} />;
    }
  };

const PageScaffold = () => (…);

export default withPageViews(PageScaffold);
```

**Comment by trezy** [comment]
For anybody coming late to the party, the reason @kylewiedman's solution doesn't work is because the Router events are being defined inside of `componentDidMount`. Defining them outside of the component seems to fix the problem:

https://gist.github.com/trezy/e26cb7feb2349f585d2daf449411d0a4

**Comment by kachkaev** [comment]
@trezy when using `gtag`, there seems to be no need to `import ReactGA from 'react-ga'`. 

I used your `<script dangerouslySetInnerHtml={...} />` example and just added this one line into the injection:
```js
window.gaTrackingId = '${gaTrackingId}';
```

Afterwards, this lightweight trick began to work:

```js
Router.onRouteChangeComplete = () => {
  NProgress.done();
  if (window.gtag) {
    window.gtag('config', window.gaTrackingId, {
      page_location: window.location.href,
      page_path: window.location.pathname,
      page_title: window.document.title,
    });
  }
};
```

**Comment by nmaro** [comment]
Since this is not obvious, it would be good to have an example in the repo.

**Comment by timneutkens** [comment]
Or a next-google-analytics repository 🕵️

**Comment by Vadorequest** [comment]
Anyone with a bit of experience would be kind to clarify the GA/GA-old/GTM mix for me?

I've spent the afternoon working on https://github.com/Vadorequest/react-gtm (because I couldn't find a package that fit my needs, so I forked and made this one)

I am using GTM (Google Tag Manager) and I basically inject this in the DOM on page load:

```js
  const scriptTag = `
    (function(w,d,s,l,i){w[l]=w[l]||[];w[l].push({'gtm.start':
      new Date().getTime(),event:'gtm.js',${JSON.stringify(events).slice(1, -1)}});
      var f=d.getElementsByTagName(s)[0],j=d.createElement(s),dl=l!='dataLayer'?'&l='+l:'';
      j.async=true;j.src='${protocol}://www.googletagmanager.com/gtm.js?id='+i+dl+'${gtm_auth}${gtm_preview}&gtm_cookies_win=x';
      f.parentNode.insertBefore(j,f);
    })(window,document,'script','${dataLayerName}','${id}');`;
```

I have set up a Universal Google Analytics on my GTM, and I have `ga` available in the console. I have sucessfully used the `dataLayer` feature of GTM, but I don't see anything happening on GA console. (except for the fact that I there is 1 active user, but no pageview at all)

I thought using GTM dataLayer/events was related to GA, I don't quite understand what is the use of GTM dataLayer/events if they aren't storing anything into GA. (I'm using the new Environments feature of GTM, and I'm basically working on my staging environment, maybe no data are shown in GA because of that? I don't know)

Also, I'm just figuring out that the page doesn't get tracked on the client side (works fine for server-side redirections), if I refresh the page, I can see my initial dataLayer, and then I see my actions triggering my events. But I have no idea where all this gets stored, I just see it because I enabled the preview/debug mode of GTM.

And now, I hear about the "new GA way", which relies on Gtag... I'm getting lost.

So:

1. Do I need to track pageviews using GA, even though I included GA from GTM? (It appears so)
1. Where are stored GTM dataLayer and events data? Is it even useful to use them?
1. Should I use gtag instead of ga?

**Comment by guigrpa** [comment]
@prichodko Thanks a lot for putting together this example!!

**Comment by andylacko** [comment]
found pretty nice lib, added to main layout file in componentDidMount() and it seems to work correctly. It is well documented, no hustle

[https://www.npmjs.com/package/react-gtm-module](https://www.npmjs.com/package/react-gtm-module)

**Comment by williamli** [comment]
Just adding my solution to this here in case it might be useful for someone later on.

I used [react-ga](https://github.com/react-ga/react-ga) but the same logic should work with other ga tools.

`_app.js`'s getInitialProps is triggered on the client side whenever user click on a new route. (First time it will run on server side).

`componentDidMount` of `_app.js` runs on client side only.

So in my `_app.js`,  I added the following few lines of code

```

static async getInitialProps({ Component, router, ctx }) {
    let pageProps = {}

    if (Component.getInitialProps) {
      pageProps = await Component.getInitialProps(ctx);      
    }

    // client-side only, run on page changes, do not run on server (SSR)
    if (typeof(window) === undefined) {
      ReactGA.pageview(router.asPath);
    }

    return { pageProps, router }
  }
  
  
  componentDidMount() {
    // client-side only, run once on mount
    ReactGA.initialize('UA-XXXXXXX-3');
    ReactGA.pageview(window.location.pathname + window.location.search);
  }

```


When the page is rendered in server, nothing will happen in `getInitialProps` as I wrapped the GA code in `typeof(window) === undefined`.

On client side, `getInitialProps` will not run for the first time since everything is server rendered. GA is setup inside `componentDidMount` on the client side.

Subsequent route changes (even for the same route), will trigger `getInitialProps` on client side and ga event is triggered.

**Comment by ChrisEdson** [comment]
> Just adding my solution to this here in case it might be useful for someone later on.
> 
> I used [react-ga](https://github.com/react-ga/react-ga) but the same logic should work with other ga tools.
> 
> `_app.js`'s getInitialProps is triggered on the client side whenever user click on a new route. (First time it will run on server side).
> 
> `componentDidMount` of `_app.js` runs on client side only.
> 
> So in my `_app.js`, I added the following few lines of code
> 
> ```
> 
> static async getInitialProps({ Component, router, ctx }) {
>     let pageProps = {}
> 
>     if (Component.getInitialProps) {
>       pageProps = await Component.getInitialProps(ctx);      
>     }
> 
>     // client-side only, run on page changes, do not run on server (SSR)
>     if (typeof(window) === "object") {
>       ReactGA.pageview(ctx.asPath);
>     }
>     return { pageProps, router }
>   }
>   
>   
>   componentDidMount() {
>     // client-side only, run once on mount
>     ReactGA.initialize('UA-XXXXXXX-3');
>     ReactGA.pageview(window.location.pathname + window.location.search);
>   }
> ```
> When the page is rendered in server, nothing will happen in `getInitialProps` as I wrapped the GA code in `typeof(window) === "object"`.
> 
> On client side, `getInitialProps` will not run for the first time since everything is server rendered. GA is setup inside `componentDidMount` on the client side.
> 
> Subsequent route changes (even for the same route), will trigger `getInitialProps` on client side and ga event is triggered.

Hey Will. Does including ReactGA in your webpack (as it's run clientside) have a significant effect on your build size?

**Comment by williamli** [comment]
@ChrisEdson https://github.com/react-ga/react-ga the whole library is 161kb zipped.

**Comment by ChrisEdson** [comment]
Hmm. That’s absolutely huge to be honest!

**Comment by williamli** [comment]
that’s the number listed on GitHub for the whole library. 
How do I check the actual size once packaged in app?

**Comment by ChrisEdson** [comment]
So the minified script is 15kb. That's quite small.

I'm going to bundle analyse my webpack and I'll report back on how much it adds in practice.

**Comment by ChrisEdson** [comment]
Within a total bundle size of ~500kb, my ReactGA adds about 27kb. So pretty small in the grand scheme of things.

**Comment by tunelinks** [comment]
> Official example available here: https://github.com/zeit/next.js/tree/canary/examples/with-google-analytics

Can you guys add page title to this example please.  I only see pageview (url)

**Comment by tunelinks** [comment]
Has anyone run into the issue where the Page Title and Page URL are different in Real Time Analytics?

**Comment by GodOfGrandeur** [comment]
> Just adding my solution to this here in case it might be useful for someone later on.
> 
> I used [react-ga](https://github.com/react-ga/react-ga) but the same logic should work with other ga tools.
> 
> `_app.js`'s getInitialProps is triggered on the client side whenever user click on a new route. (First time it will run on server side).
> 
> `componentDidMount` of `_app.js` runs on client side only.
> 
> So in my `_app.js`, I added the following few lines of code
> 
> ```
> 
> static async getInitialProps({ Component, router, ctx }) {
>     let pageProps = {}
> 
>     if (Component.getInitialProps) {
>       pageProps = await Component.getInitialProps(ctx);      
>     }
> 
>     // client-side only, run on page changes, do not run on server (SSR)
>     if (typeof(window) === "object") {
>       ReactGA.pageview(ctx.asPath);
>     }
>     return { pageProps, router }
>   }
>   
>   
>   componentDidMount() {
>     // client-side only, run once on mount
>     ReactGA.initialize('UA-XXXXXXX-3');
>     ReactGA.pageview(window.location.pathname + window.location.search);
>   }
> ```
> 
> When the page is rendered in server, nothing will happen in `getInitialProps` as I wrapped the GA code in `typeof(window) === "object"`.
> 
> On client side, `getInitialProps` will not run for the first time since everything is server rendered. GA is setup inside `componentDidMount` on the client side.
> 
> Subsequent route changes (even for the same route), will trigger `getInitialProps` on client side and ga event is triggered.

Is there a better place to put the `ReactGA.pageview(ctx.asPath);`, removing it from `getInitialProps` mean we can remove `getInitialProps` and not force Automatic Prerendering to opt-out as outlined here: https://err.sh/next.js/opt-out-automatic-prerendering

**Comment by hkirsman** [comment]
@GodOfGrandeur This sounds promissing, will try in prod. Didn't get the version mentioned above working or seemed like so https://medium.com/@austintoddj/using-google-analytics-with-next-js-423ea2d16a98 Adding your code in _app.js seems to trigger everywhere like needed and I also like it's not triggered on server (would not have though it's an issue).

**Comment by elhakeem** [comment]
No need for any third party, If you are using gtag use the following code I created:-
```
<script async src="https://www.googletagmanager.com/gtag/js?id=<%your code here%>"></script>
    <script dangerouslySetInnerHTML={{__html: `
    window.dataLayer = window.dataLayer || [];
    function gtag(){window.dataLayer.push(arguments)}
    gtag("js", new Date());
    gtag("config", "<%your code here%>");
    `}}>
    </script>
```

**Comment by reinink** [comment]
Another option is to use `useEffect()`. I've done this within my `Layout.js` file.

```js
useEffect(() => {
  if (process.env.NODE_ENV === 'production') {
    window.dataLayer = window.dataLayer || []
    function gtag() {
      dataLayer.push(arguments)
    }
    gtag('js', new Date())
    gtag('config', '{GOOGLE ANALYTICS CODE}', {
      page_location: window.location.href,
      page_path: window.location.pathname,
      page_title: window.document.title,
    })
  }
})
```

And in the `<Head>`:

```html
<Head>
  <script async src="https://www.googletagmanager.com/gtag/js?id={GOOGLE ANALYTICS CODE}"></script>
</Head>
```

**Comment by amorriscode** [comment]
@reinink Just want to mention that you might want to pass an empty array as the second prop to `useEffect` so that it only runs the function once after the initial render.

```JavaScript
useEffect(() => {...}, [])
```

**Comment by antonmyrberg** [comment]
Expanding on @reinink s solution, here's a working HOC I use to wrap my layout components/templates. 

```javascript
import React, { useEffect } from 'react'
import Head from 'next/head'
const GoogleAnalyticsHOC = ({ children }) => {
  useEffect(() => {
    if (process.env.NODE_ENV === 'production') {
      window.dataLayer = window.dataLayer || []
      // eslint-disable-next-line
      function gtag() {
        window.dataLayer.push(arguments)
      }
      gtag('js', new Date())
      gtag('config', process.env.GOOGLE_ANALYTICS_ID, {
        page_location: window.location.href,
        page_path: window.location.pathname,
        page_title: window.document.title
      })
    }
  }, [])
  return (
    <>
      <Head>
        <script async src={`https://www.googletagmanager.com/gtag/js?id=${process.env.GOOGLE_ANALYTICS_ID}`} />
      </Head>
      {children}
    </>
  )
}
export default GoogleAnalyticsHOC
```

It's not extensively tested but it works and passes myself as a realtime visitor running locally. :)

**Comment by iaurg** [comment]
Hi folks, I had the same issue and solved using Google Tag manager (plus: I can put any script via tag manager):

In you _app.js
```
import TagManager from 'react-gtm-module';
import Router from 'next/router';

/*With this router my GA can catch the page change in app*/
Router.events.on('routeChangeComplete', () => {
  ga('gtag_YOUR-TAG-ID.send', 'pageview', location.pathname);
});

const tagManagerArgs = {
  gtmId: 'GTM-YOURID',
  auth: 'YOUR-GTM-AUTH',
};

class MyApp extends App {
  componentDidMount() {
    TagManager.initialize(tagManagerArgs);
  }
```
If your know better approach please comment.

**Comment by UberVero** [comment]
I found [this guide](https://coderrocketfuel.com/article/add-google-analytics-to-a-next-js-and-react-website) - is this a recommended approach (uses react-ga)?

**Comment by JerryGreen** [comment]
I'd say recommended way is to use this official example:

https://github.com/zeit/next.js/tree/canary/examples/with-google-analytics

It's already posted here. Sometimes I wish github issues to be more like stackoverflow, having some answer accepted and highlighted, right after the question (when there's too many messages and emojis is not enough. One option is to restrict commenting this post, so people will create a new issue instead. And while creating they'll find a similar issue was already created (thanks to "similar issues" feature), and it has an answer (but still, the answer with emojis is lost somewhere in the middle of long history, so it's not very convenient to find answers on github).

**Comment by janbaykara** [comment]
Github is prototyping this over on this repo's [Discussions](https://github.com/zeit/next.js/discussions/) tab.

<img width="788" alt="Screenshot 2020-04-08 at 16 59 58" src="https://user-images.githubusercontent.com/237556/78806305-013aed00-79b2-11ea-9620-7025a7c37898.png">

**Comment by JerryGreen** [comment]
@janbaykara wow, really? We all think the same way :) I hope they will find a way to somehow translate an issue to a discussion though, or it will be the same thing (if functionality will be merged to issues, and all existing discussions will become issues). But thanks for the response!

**Comment by hpachy** [comment]
Hello ! 

I got the problem why the gtag implementation. First of all the Gtag does not work (obvious) and my complilator throw me 

> Error: Minified React error #321  (more informations => https://stackoverflow.com/questions/62026031/useeffect-error-minified-react-error-321)

As I explain in the stack overFlow thread I tried to do the same example as Zeit with typeScript

`
// _app.ts
import { useEffect } from 'react';
import Router from 'next/router';

import React, { ReactElement } from 'react';

import { AppProps } from 'next/app';
import '../CSS/index.css';
import * as gtag from '../lib/gtag';
import '@brainhubeu/react-carousel/lib/style.css';

export default function  App({ Component, pageProps }: AppProps): ReactElement {
	useEffect(() => {
		const handleRouteChange = (url: string) => {
			gtag.pageview(url);
		};
		Router.events.on('routeChangeComplete', handleRouteChange);
		return () => {
			Router.events.off('routeChangeComplete', handleRouteChange);
		};
	}, []);

	return <Component {...pageProps} />;
}
`

At the 'next build' that's a part of error throwed 
`
Error occurred prerendering page "/***/***". Read more: https://err.sh/next.js/prerender-error
Error: Minified React error #321; visit https://reactjs.org/docs/error-decoder.html?invariant=321 for the full message or use the non-minified dev environment for full errors and additional helpful warnings.
Automatically optimizing pages ..Error: Minified React error #321; visit https://reactjs.org/docs/error-decoder.html?invariant=321 for the full message or use the non-minified dev environment for full errors and additional helpful warnings.
    at Z (/Users/***/***/***/.next/serverless/pages/_error.js:6820:404)
    at module.exports.viRO.exports.useEffect (/Users/***/***/***/.next/serverless/pages/_error.js:6825:356)
    at App (/Users/***/***/***/.next/serverless/pages/_error.js:4238:58)
    at d (/Users/***/***/***/.next/serverless/pages/_error.js:694:498)
    at $a (/Users/***/***/***/next/serverless/pages/_error.js:697:16)
    at a.b.render (/Users/***/***/***/.next/serverless/pages/_error.js:702:476)
    at a.b.read (/Users/***/***/***/.next/serverless/pages/_error.js:702:18)
    at renderToString (/Users/***/***/***/.next/serverless/pages/_error.js:712:364)
    at render (/Users/***/***/***/.next/serverless/pages/_error.js:248:16)
    at Object.renderPage (/Users/***/***/***/.next/serverless/pages/_error.js:570:16)

Error occurred prerendering page "/404". Read more: https://err.sh/next.js/prerender-error
Error: Minified React error #321; visit https://reactjs.org/docs/error-decoder.html?invariant=321 for the full message or use the non-minified dev environment for full errors and additional helpful warnings.

> Build error occurred
Error: Export encountered errors
    at _default (/usr/local/lib/node_modules/next/dist/export/index.js:19:1164)
    at process._tickCallback (internal/process/next_tick.js:68:7)
Automatically optimizing pages 
`
I would like to understand what wrong... can someone help me to understand what I missed ?

**Comment by 3lonious** [comment]
to be honest ive tried all this stuff and nothing is working, the next js example is pretty confusing if you have an already built app/site... seriously why is this so hard to add a dam script tag of any kind? seriously thinking of ditching next js because ive spent way to much time trying to figure this out and its driving me crazy... can we see a full example with detailed instructions on how to add ANY SCRIPT TAG period to a next js app? WTF why is this not fixed? i have many script tags from different companies on my site that i need to be working and pulling my hair out to even get googles scripts working, seriously, how is this even a thing

**Comment by Vadorequest** [comment]
@3lonious I've personally ditched Google Analytics, as it does not do a good job with SPA in general. (really bad experience, especially with SSR+CSR and double events hard to track down)

Regarding Next.js, check out https://github.com/UnlyEd/next-right-now/blob/v2-mst-aptd-gcms-lcz-sty/src/components/pageLayouts/Head.tsx (using `next/Head`), that's how I include script tags in my apps.

Dont hesitate to take a look at NRN https://unlyed.github.io/next-right-now, it's meant to be a good example on "how to do X" with Next.js and should save you quite some headaches.

**Comment by 3lonious** [comment]
im not sure i get it, i dont see any script tags in that file?

**Comment by antonmyrberg** [comment]
This is also a solution that works. I haven't tested it extensively so please do that yourself.

In `pages/_document.js`, add this to your `<Head>`. See the docs on [how to override the document file](https://nextjs.org/docs/advanced-features/custom-document).

```javascript
<script async src='https://www.googletagmanager.com/gtag/js?id=YOUR_GA_TRACKING_ID'></script>
<script
    dangerouslySetInnerHTML={{
        __html: `
            window.dataLayer = window.dataLayer || [];
            function gtag(){dataLayer.push(arguments)}
            gtag('js', new Date());
            
            gtag('config', 'YOUR_GA_TRACKING_ID');
              `
      }}>
</script>
```

It reports traffic to my GA account at least.

**Comment by Vadorequest** [comment]
@3lonious Script tags and style tags work the same way.
The above solution also works, from experience. (it's how I used to make GA work)

**Comment by 3lonious** [comment]
ok thanks guys. im gonna try and get back to you

**Comment by 3lonious** [comment]
ok well this is the tag i got, what do you do with that?


<script>(function(w,d,s,l,i){w[l]=w[l]||[];w[l].push({'gtm.start':
new Date().getTime(),event:'gtm.js'});var f=d.getElementsByTagName(s)[0],
j=d.createElement(s),dl=l!='dataLayer'?'&l='+l:'';j.async=true;j.src=
'https://www.googletagmanager.com/gtm.js?id='+i+dl;f.parentNode.insertBefore(j,f);
})(window,document,'script','dataLayer','SOME_ACCOUNT_NUMS');</script>

**Comment by 3lonious** [comment]
ok i think i got it to work

**Comment by 3lonious** [comment]
thanks Anton. ok i have a new one for you haha

i have a chat app i use for customer service but that script tag dosnt work either? any ideas?

**Comment by 3lonious** [comment]
its telling me to add this second tag in the body? what do you do with that?
<img width="441" alt="Screen Shot 2020-06-19 at 6 48 22 AM" src="https://user-images.githubusercontent.com/46793692/85139190-ecbc7180-b1f8-11ea-8ec3-a79c596f1247.png">

**Comment by 3lonious** [comment]
so actually it looks like this script i added yesterday is now working and im not sure why.. looks like the script from Anton actually wasn't working.. weird

 <script async src='https://www.googletagmanager.com/gtag/js?id=ACCOUNT_ID'></script>

**Comment by ernaneluis** [comment]
How About TBT? If you add google analytics it will slower your speed score

**Comment by JerryGreen** [comment]
> How About TBT? If you add google analytics it will slower your speed score

That’s because of their recommendation of placing the script before the content. If you place it under the content, the score won’t go down.

I personally don’t see real benefits of placing the script first. If the page load was somehow broken, and HTML didn’t fully load somehow and it was aborted, then I might not count it as a hit, so whatever.

**Comment by ChaitraK2000** [comment]
I m working with nextjs. I m  trying to use react google place auto complete API. And wants to insert those script tag of API? Where should I insert those script tag??

**Comment by Poshtarenko-Dmitriy** [comment]
**_app.js** 

```js
    import ReactGA from "react-ga";
 
    componentDidMount() {
    ReactGA.initialize("XX-XXXXXXXXX-X", { debug: false });
    ReactGA.set({ page: this.props.router.pathname });
    ReactGA.pageview(this.props.router.pathname);
    this.unlisten = this.props.router.events.on("routeChangeComplete", (router) => {
        ReactGA.set({ page: router });
        ReactGA.pageview(router);
      });
    }

**Comment by jfbrennan** [comment]
This is hilarious. No, actually this is sad. Can't wait for React, Next, the whole ecosystem to die.

**Comment by RiusmaX** [comment]
With React hooks : 

**_app.js**

` import { useEffect } from 'react'
import { useRouter } from 'next/router'
import ReactGA from 'react-ga'
import Layout from '../components/Layout'

function MyApp ({ Component, pageProps }) {
  const router = useRouter()

  useEffect(() => {
    const handleRouteChange = (url, { shallow }) => {
      ReactGA.set({ page: url })
      ReactGA.pageview(url)
    }
    ReactGA.initialize('XX-XXXXXXXXX-X', { debug: false })
    ReactGA.set({ page: router.pathname })
    ReactGA.pageview(router.pathname)
    router.events.on('routeChangeComplete', handleRouteChange)
    return () => {
      router.events.off('routeChangeComplete', handleRouteChange)
    }
  }, [])

  return (
    <Layout>
      <Component {...pageProps} />
    </Layout>
  )
}

export default MyApp`

**Comment by balazsorban44** [comment]
This issue has been automatically locked due to no recent activity. If you are running into a similar issue, please create a new issue with the steps to reproduce. Thank you.
